pysph/tools/ipython.py

```diff
@@ -0,0 +1,590 @@
+                    self._widgets.particles[particles_type].vector_width,
+                    self._widgets.particles[particles_type].scalar_size,
+                    self._widgets.particles[particles_type].legend,
+
+                ],
+                    layout=Layout(display='flex'))
+            )
+
+        display(
+            VBox(
+                [
+                    HBox(items, layout=Layout(display='flex')),
+                    self._widgets.frame,
+                    self._widgets.save_figure
+                ]
+            )
+        )
+
+    def _configure_plot(self):
+        '''
+        Set attributes for plotting.
+        '''
+
+        from matplotlib import pyplot as plt
+
+        self.figure = plt.figure()
+        self._scatter_ax = self.figure.add_axes([0, 0, 1, 1])
+        self._scatters = {}
+        self._cbar_ax = {}
+        self._cbars = {}
+        self._vectors = {}
+
+    def interactive_plot(self):
+        '''
+        Set plotting attributes, create widgets and display them
+        along with the interactive plot.
+
+        Use %matplotlib notebook for more interactivity.
+        '''
+
+        self._configure_plot()
+        self._create_widgets()
+        self._frame_handler(None)
+        self._display_widgets()
+
+    def _frame_handler(self, change):
+
+        temp_data = self.get_frame(self._widgets.frame.value)
+
+        # scalars #
+
+        self.figure.set_label(
+            "Time : " + str(temp_data['solver_data']['t'])
+        )
+        temp_data = temp_data['arrays']
+
+        for sct in self._scatters.values():
+            if sct in self._scatter_ax.collections:
+                self._scatter_ax.collections.remove(sct)
+
+        self._scatters = {}
+        for particles_type in self._widgets.particles.keys():
+            if self._widgets.particles[particles_type].scalar.value != 'None':
+                self._scatters[particles_type] = self._scatter_ax.scatter(
+                    temp_data[particles_type].x,
+                    temp_data[particles_type].y,
+                    c=getattr(
+                            temp_data[particles_type],
+                            self._widgets.particles[
+                                particles_type
+                            ].scalar.value
+                    ),
+                    s=self._widgets.particles[
+                        particles_type
+                    ].scalar_size.value,
+                )
+        self._legend_handler(None, manual=True)
+
+        # _vectors #
+
+        for vct in self._vectors.values():
+            if vct in self._scatter_ax.collections:
+                self._scatter_ax.collections.remove(vct)
+
+        self._vectors = {}
+        for particles_type in self._widgets.particles.keys():
+            if self._widgets.particles[particles_type].vector.value != '':
+                self._vectors[particles_type] = self._scatter_ax.quiver(
```

Comment:
Again same pattern, repeated `temp_data[particles_type]` repeated `self._widgets.particle[particle_type]` etc.  You can make the code much easier to read and write than this.